### PR TITLE
Add Activity Links support

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Adding support for exporting activity links as _MS.links

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/TelemetryPartB.cs [hunk 1]
```diff
@@ -123,6 +129,58 @@ private static void AddPropertiesToTelemetry(IDictionary<string, string> destina
...
+                        .Append(link.Context.SpanId.ToHexString())
+                        .Append('\"');
+                    linksJson.Append("},");
+
+                    if (linksJson.Length > maxLength)
+                    {
+                        break;
+                    }
+
+                    prevIndex = linksJson.Length;
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/TelemetryPartB.cs:163]: Getting a length of stringbulider every time is costlier. We could decrement the constant link length from 8192 and looks out for its length is greater than 0.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/TelemetryPartB.cs [hunk 2]
```diff
@@ -123,6 +129,53 @@ private static void AddPropertiesToTelemetry(IDictionary<string, string> destina
             }
         }
 
+        private static void AddActivityLinksToPartCTags(IEnumerable<ActivityLink> links, ref AzMonList PartCTags)
+        {
+            string msLinks = "_MS.links";
+            // max number of links that can fit in this json formatted string is 107. it is based on assumption that traceid and spanid will be of fixed length.
+            // Keeping max at 100 for now.
+            // todo: define this as constant.
+            int maxLinksAllowed =100;
```
Comment (vishweshbankwar) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/TelemetryPartB.cs:138]: I kept it 100 here for now. Max links that can fit into 8192 character limit are 107. Open for suggestions/comments.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/TelemetryPartB.cs [hunk 3]
```diff
@@ -123,6 +129,53 @@ private static void AddPropertiesToTelemetry(IDictionary<string, string> destina
...
+                        .Append("\"id\":")
+                        .Append('\"')
+                        .Append(link.Context.SpanId.ToHexString())
+                        .Append('\"');
+                    linksJson.Append("},");
+
+                    maxLinksAllowed--;
+                    if (maxLinksAllowed == 0)
+                    {
+                        break;
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/TelemetryPartB.cs:163]: some logs to indicate we are stopping to look for links, and there are more links.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/TelemetryPartB.cs [hunk 4]
```diff
@@ -123,6 +129,53 @@ private static void AddPropertiesToTelemetry(IDictionary<string, string> destina
             }
         }
 
+        private static void AddActivityLinksToPartCTags(IEnumerable<ActivityLink> links, ref AzMonList PartCTags)
+        {
+            string msLinks = "_MS.links";
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/TelemetryPartB.cs:134]: could you add a code comment showing the sample json being built here.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/TelemetryPartCTests.cs [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+            using var activity = activitySource.StartActivity(
+                ActivityName,
+                ActivityKind.Client,
+                parentContext: default,
+                null,
+                links,
+                startTime: DateTime.UtcNow);
+
+            // Only checking dependency here. The flow will be same for other telemetry types.
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/TelemetryPartCTests.cs:146]: you can parameterize the test and check both types, without duplicating a lot of code.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/TelemetryPartCTests.cs [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+        }
+
+        [Fact]
+        public void LinksAreTruncatedWhenCannotFitInMaxLength()
+        {
+            using ActivitySource activitySource = new ActivitySource(ActivitySourceName);
+            List<ActivityLink> links = new List<ActivityLink>();
+
+            int numberOfLinks = 150; //arbitrary number > 100
+            int maxLinksAllowed = 100;
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/TelemetryPartCTests.cs:124]: don't define them define. refer from a common place.

**Review by reyang — APPROVED** [review]

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/TelemetryPartCTests.cs (hunk 6 above)
Comment (reyang) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/TelemetryPartCTests.cs]: maybe make it internal from the exporter, and make the exporter InternalsVisibleTo the tests.

Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/TelemetryPartCTests.cs]: Yes - I am going to find a common place for all such constraints when I add the complete truncation logic. Will move it then

**Review by rajkumar-rangaraj — APPROVED** [review]

**Pull request merged**